src/storage/conversion/MarkdownToHtmlConverter.ts

```diff
@@ -0,0 +1,37 @@
+import { promises as fsPromises } from 'fs';
+import marked from 'marked';
+import sanitizeHtml from 'sanitize-html';
+import { BasicRepresentation } from '../../ldp/representation/BasicRepresentation';
+import type { Representation } from '../../ldp/representation/Representation';
+import type { TemplateEngine } from '../../pods/generate/TemplateEngine';
+import { TEXT_HTML, TEXT_MARKDOWN } from '../../util/ContentTypes';
+import { resolveAssetPath } from '../../util/PathUtil';
+import { readableToString } from '../../util/StreamUtil';
+import type { RepresentationConverterArgs } from './RepresentationConverter';
+import { TypedRepresentationConverter } from './TypedRepresentationConverter';
+
+/**
+ * Converts markdown data to HTML.
+ * The generated HTML will be injected into the given template using the parameter `body`.
+ * A standard markdown string will be converted to a <p> tag, so html and body tags should be part of the template.
+ */
+export class MarkdownToHtmlConverter extends TypedRepresentationConverter {
+  private readonly engine: TemplateEngine;
+  private readonly templatePath: string;
+
+  public constructor(engine: TemplateEngine, templatePath: string) {
+    super(TEXT_MARKDOWN, TEXT_HTML);
+    this.engine = engine;
+    this.templatePath = resolveAssetPath(templatePath);
+  }
+
+  public async handle({ representation }: RepresentationConverterArgs): Promise<Representation> {
+    const markdown = await readableToString(representation.data);
+    const body = sanitizeHtml(marked(markdown));
```

Comment:
Why is the sanitization needed?

Reply:
Mostly because it was suggested in the npm library and I was like, "why not". The converter does not know anything about the markdown it receives so it seemed like not a bad idea.

Reply:
Yeah but sanitization is a heuristic; better to have the input right I think. Better to know the markdown is untrusted than to rely on heuristics that might or might not be trusted.

Reply:
I just remembered that this is actually the reason we don't send error message in case an error happens after headers were sent, since that was a potential point where something could be injected in the error message https://github.com/solid/community-server/blob/dfb8ac938d96291c7bddfd79cc4dad39af213a35/src/server/BaseHttpServerFactory.ts#L78-L83

Didn't take this into account when writing the html error generator. It's also just easier to do it there. If we want to clean it at the point of origin we would have to do it every time creating an error message that contains a reference to something that was input data.

It might not be the perfect solution to have it here, but the advantage is that we can reduce it to this single place though.

Reply:
Our HTML is already sanitized because values are escaped by the Handlebars engine.

Any unsanitized HTML is due to faulty templates, and there is no waterproof fix for that.

Reply:
> Our HTML is already sanitized because values are escaped by the Handlebars engine.

That's actually the part I disabled 😅 . In total we can generate up to 3 templates that need to be injected into a larger template. If handlebars sanitized those values the templates would not work anymore. The first template is the error description (E0001), which gets injected into the main error markdown template, which gets converted into html and injected into the main html template.

For the first one, handlebars removed some of the tick marks when sanitizing the input. Also the result is markdown anyway so it doesn't make sense to apply sanitization there of another content-type (html).

When converting the markdown to html the output is something like `<p>Error message</p>` which needs to be injected into the main html template, which again would not work at all if handlebars was allowed to sanitize.

Reply:
Oh no no, that's not good.

Use triple-slash when you expect HTML (https://handlebarsjs.com/guide/expressions.html#html-escaping) and name your variables `descriptionHtml` or whatever to emphasize that.

Reply:
Basically we're open to injection now. Sanitize will remove invalid HTML, but not nicely valid injected scripts.

Reply:
Yea by disabling it I meant using the triple braces, not disabling it in the handlebars config. It's actually true that the first time one of the variables is injected in the template it's done with the normal double braces. It's only when the body it is in afterwards is injected again it's done with triple braces so it's fine in those cases.

I did also add triple braces to the error message and stack just in case they could have something ruined by handlebars, but in hindsight that's not really necessary and those can be reverted to double braces. So in that case all injected values pass through double braces at least once.

This does make us depend on the templating engine being used having something like that built in, but it is the easiest solution.

Reply:
Exactly; we trust our own templates to be good HTML but not the error messages.

Reply:
I'd still suggest the `...html` naming though for safety.